File: ios/WebyarNative/Sources/Features/Calls/CallSession.swift
import Foundation
import AVFoundation
import Observation
import LiveKit

/// Where a call is in its life.
///
/// One value rather than a set of booleans, for the same reason `LoadState`
/// exists: it makes "connecting and also ended" unrepresentable, and every
/// one of those combinations is a bug somebody would otherwise have to find.
enum CallPhase: Equatable, Sendable {
    /// The invitation is out; the visitor has not answered.
    case waiting
    /// They answered; we are joining the room.
    case connecting
    /// Both sides are in.
    case connected
    /// Over, for the stated reason.
    case ended(CallOutcome)

    var isLive: Bool {
        switch self {
        case .waiting, .connecting, .connected: true
        case .ended: false
        }
    }
}

/// A call that connected, but not with everything it was supposed to carry.
///
/// Worth its own type rather than a boolean: an operator who cannot be heard
/// and an operator who cannot be seen have different problems, and telling
/// them "something went wrong" helps with neither.
enum CallDegradation: Equatable, Sendable {
    /// The microphone would not start — permission, or another app holding it.
    case noMicrophone
    /// The camera would not start. The call carries on as audio.
    case noCamera

    func title(_ language: Language) -> String {
        switch self {
        case .noMicrophone: Str.callNoMicrophone(language)
        case .noCamera: Str.callNoCamera(language)
        }
    }
}

/// How a call finished. Each of these reads differently to an operator, and
/// telling them apart is most of what makes the screen feel honest.
enum CallOutcome: Equatable, Sendable {
    /// The operator hung up.
    case hungUp
    /// The visitor hung up, or closed the page.
    case visitorLeft
    /// The visitor said no.
    case declined
    /// Nobody answered in time.
    case expired
    /// We never got into the room.
    case failed(String)
}

/// One call, from the invitation going out to the room closing.
///
/// The shape of this deliberately mirrors what the operator console does,
/// because the two have to agree: the same invitation is polled, the same
/// `POST /api/calls/:id/token` is minted, and the same LiveKit room is joined
/// with the same participant type. A phone that took a shortcut here would be
/// a phone whose calls behave subtly differently from the desktop's, and
/// nobody would be able to say how.
@MainActor
@Observable
final class CallSession {

    private(set) var phase: CallPhase = .waiting
    /// The visitor, once they are actually in the room.
    private(set) var visitorPresent = false
    private(set) var isMuted = false
    private(set) var isCameraOn: Bool
    private(set) var isSpeakerOn = true
    /// When the two sides actually connected, which is what the timer counts
    /// from — not when the invitation was sent.
    private(set) var connectedAt: Date?
    /// Set when the server warned that TURN is unconfigured. Not fatal, but
    /// the first thing to look at when a call fails on a mobile network.
    private(set) var relayWarning = false
    /// Set when the call connected but something could not be published.
    private(set) var degraded: CallDegradation?

    let channel: CallChannel
    let contactName: String
    let contactAvatarURL: String?

    /// The room, once there is one. `nil` outside a connected call, which is
    /// also what releases the microphone.
    private(set) var room: Room?

    private let invitationID: String
    private let api: any WebyarAPI
    private var callSessionID: String?
    private var pollTask: Task<Void, Never>?
    private var roomDelegate: RoomObserver?

    init(
        invitation: CallInvitation,
        contactName: String,
        contactAvatarURL: String?,
        api: any WebyarAPI = Backend.current
    ) {
        self.invitationID = invitation.id
        self.channel = invitation.kind
        self.isCameraOn = invitation.kind == .video
        self.contactName = contactName
        self.contactAvatarURL = contactAvatarURL
        self.api = api
    }

    // MARK: - Lifecycle

    /// Waits for the visitor, then joins.
    ///
    /// Polling rather than realtime on purpose: the app has no realtime
    /// transport yet, an invitation lives for five minutes, and a request
    /// every two seconds for at most that long is cheaper to build and to
    /// reason about than a socket that exists for this one screen.
    func start() {
        guard pollTask == nil else { return }
        pollTask = Task { [weak self] in
            await self?.waitForVisitor()
        }
    }

    private func waitForVisitor() async {
        while !Task.isCancelled, phase == .waiting {
            do {
                let invitation = try await api.invitation(id: invitationID)
                if let sessionID = invitation.callSessionId, invitation.isJoined {
                    callSessionID = sessionID
                    await join(callSessionID: sessionID)
                    return
                }
                if invitation.isTerminal {
                    phase = .ended(invitation.status == "declined" ? .declined : .expired)
                    return
                }
            } catch {
                // A single failed poll is a blip, not an answer. Only a
                // terminal invitation or the operator ends the wait.
            }
            try? await Task.sleep(for: .seconds(2))
        }
    }

    private func join(callSessionID: String) async {
        phase = .connecting
        do {
            let credentials = try await api.callToken(
                callSessionID: callSessionID,
                displayName: nil
            )
            relayWarning = credentials.warnings?.contains("turn_missing") == true

            guard let url = credentials.signallingURL else {
                phase = .ended(.failed("no_server_url"))
                return
            }

            // No audio-session setup here: `AudioManager` inside the SDK owns
            // the session and configures `.playAndRecord` itself. Doing it by
            // hand is what made the engine tear down and rebuild mid-call.
            let room = Room()
            let observer = RoomObserver(session: self)
            room.add(delegate: observer)
            roomDelegate = observer
            self.room = room

            try await room.connect(
                url: url,
                token: credentials.token,
                connectOptions: ConnectOptions(
                    iceServers: iceServers(from: credentials),
                    iceTransportPolicy: credentials.relayOnly ? .relay : .all
                ),
                roomOptions: RoomOptions(
                    defaultCameraCaptureOptions: CameraCaptureOptions(
                        position: .front,
                        dimensions: .h720_169
                    ),
                    adaptiveStream: true,
                    dynacast: true
                )
            )

            // Publishing is best-effort, deliberately. A camera that will not
            // start — no camera at all on the simulator, permission refused,
            // another app holding it — is a reason to carry on without video,
            // not a reason to drop a call the visitor has already answered.
            // Before this, one throw here took the whole call down.
            do {
                try await room.localParticipant.setMicrophone(enabled: true)
            } catch {
                isMuted = true
                degraded = .noMicrophone
            }

            if channel == .video {
                do {
                    try await room.localParticipant.setCamera(enabled: true)
                } catch {
                    isCameraOn = false
                    if degraded == nil { degraded = .noCamera }
                }
            }

            phase = .connected
            connectedAt = Date()
            refreshVisitorPresence()
        } catch {
            // Only a failure to reach the room itself gets here now.
            await finish(.failed(Self.describe(error)))
        }
    }

    /// LiveKit errors stringify into a paragraph. The operator needs the one
    /// sentence, and the log keeps the rest.
    private static func describe(_ error: Error) -> String {
        if let error = error as? LiveKitError { return error.message ?? "\(error.type)" }
        return (error as NSError).localizedDescription
    }

    private func iceServers(from credentials: CallToken) -> [IceServer] {
        guard let turn = credentials.turn, let urls = turn.urls, !urls.isEmpty else { return [] }
        return [
            IceServer(
                urls: urls,
                username: turn.username ?? "",
                credential: turn.credential ?? ""
            ),
        ]
    }

    // MARK: - Controls

    func toggleMute() {
        guard let room else { return }
        let next = !isMuted
        isMuted = next
        Task { try? await room.localParticipant.setMicrophone(enabled: !next) }
    }

    func toggleCamera() {
        guard let room, channel == .video else { return }
        let next = !isCameraOn
        isCameraOn = next
        Task { try? await room.localParticipant.setCamera(enabled: next) }
    }

    /// Earpiece or loudspeaker. A voice call starts on the loudspeaker
    /// because the operator is at a desk, not holding the phone to their ear.
    ///
    /// Routed through LiveKit rather than `AVAudioSession` directly: the SDK
    /// owns the session's configuration, and overriding the port behind its
    /// back made the audio engine tear itself down and rebuild mid-call.
    func toggleSpeaker() {
        isSpeakerOn.toggle()
        AudioManager.shared.isSpeakerOutputPreferred = isSpeakerOn
    }

    /// Ends the call for both sides.
    func hangUp() async {
        await finish(.hungUp)
    }

    /// Called when the room tells us the other side has gone.
    fileprivate func visitorDisconnected() {
        guard phase.isLive else { return }
        Task { await finish(.visitorLeft) }
    }

    /// The single way a call ends.
    ///
    /// Every path goes through here because the server has to be told on all
    /// of them — including the ones that failed. It was not, and the evidence
    /// was two call sessions left in `connecting` forever: the app had given
    /// up, the visitor's browser had not, and the platform still counted a
    /// call that nobody was on.
    private func finish(_ outcome: CallOutcome) async {
        guard phase.isLive else { return }
        phase = .ended(outcome)
        pollTask?.cancel()
        pollTask = nil

        if let callSessionID {
            // Idempotent server-side, which is what makes it safe to send
            // even when the visitor hung up first and the call is already
            // over as far as the server is concerned.
            try? await api.hangUp(callSessionID: callSessionID)
        } else {
            // Never answered: there is no session to end, only an offer to
            // withdraw.
            try? await api.cancelInvitation(id: invitationID)
        }
        await teardown()
    }

    fileprivate func refreshVisitorPresence() {
        visitorPresent = !(room?.remoteParticipants.isEmpty ?? true)
    }

    private func teardown() async {
        if let room {
            // Unpublish before disconnecting. Tearing the room down with a
            // capturer still running leaves the camera light on and makes the
            // SDK complain that it was deinitialised mid-capture.
            _ = try? await room.localParticipant.setCamera(enabled: false)
            _ = try? await room.localParticipant.setMicrophone(enabled: false)
            await room.disconnect()
        }
        room = nil
        roomDelegate = nil
    }
}

/// LiveKit talks to its delegate on its own queue; this hops each event onto
/// the main actor, where the session's state lives.
private final class RoomObserver: RoomDelegate, @unchecked Sendable {
    private weak var session: CallSession?

    init(session: CallSession) {
        self.session = session
    }

    func room(_ room: Room, participantDidConnect participant: RemoteParticipant) {
        Task { @MainActor [weak session] in session?.refreshVisitorPresence() }
    }

    func room(_ room: Room, participantDidDisconnect participant: RemoteParticipant) {
        Task { @MainActor [weak session] in
            session?.refreshVisitorPresence()
            // The visitor leaving the room ends the call: there is nobody
            // else it could be, and an operator left listening to silence
            // would have to work out for themselves that it was over.
            session?.visitorDisconnected()
        }
    }

    func room(_ room: Room, didUpdateConnectionState connectionState: ConnectionState, from oldConnectionState: ConnectionState) {
        guard connectionState == .disconnected else { return }
        Task { @MainActor [weak session] in session?.visitorDisconnected() }
    }
}

File: android/app/src/main/kotlin/com/webyar/operator/feature/call/CallSession.kt
package com.webyar.operator.feature.call

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.webyar.operator.core.model.CallChannel
import com.webyar.operator.core.model.CallInvitation
import com.webyar.operator.core.model.VisitorProfile
import com.webyar.operator.core.net.WebyarApi
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant

/**
 * One call, from the invitation going out to the room closing.
 *
 * The shape of this mirrors what the operator console does, because the two
 * have to agree: the same invitation is polled, the same
 * `POST /api/calls/:id/token` is minted, and the same room is joined. A phone
 * that took a shortcut here would be a phone whose calls behave subtly
 * differently from the desktop's, and nobody would be able to say how.
 *
 * The media is behind [CallRoom] rather than reached for directly. That is not
 * an abstraction for its own sake: WebRTC's native libraries cannot load on a
 * JVM, so a session that named them would be a session no test could
 * construct — and the parts most worth testing here are the waiting, the
 * outcomes and the degradation, none of which are about media at all.
 */
class CallSession(
    private val api: WebyarApi,
    private val room: CallRoom,
) : ViewModel() {

    private val _phase = MutableStateFlow<CallPhase>(CallPhase.Waiting)
    val phase: StateFlow<CallPhase> = _phase.asStateFlow()

    /** The visitor, once they are actually in the room. */
    private val _visitorPresent = MutableStateFlow(false)
    val visitorPresent: StateFlow<Boolean> = _visitorPresent.asStateFlow()

    private val _muted = MutableStateFlow(false)
    val muted: StateFlow<Boolean> = _muted.asStateFlow()

    private val _cameraOn = MutableStateFlow(false)
    val cameraOn: StateFlow<Boolean> = _cameraOn.asStateFlow()

    /**
     * Loudspeaker by default, even for a voice call.
     *
     * The operator is at a desk, not holding the phone to their ear.
     */
    private val _speakerOn = MutableStateFlow(true)
    val speakerOn: StateFlow<Boolean> = _speakerOn.asStateFlow()

    /**
     * When the two sides actually connected, which is what the timer counts
     * from — not when the invitation was sent.
     */
    private val _connectedAt = MutableStateFlow<Instant?>(null)
    val connectedAt: StateFlow<Instant?> = _connectedAt.asStateFlow()

    /**
     * The server warned that TURN is unconfigured.
     *
     * Not fatal, and the first thing to look at when a call fails on a mobile
     * network.
     */
    private val _relayWarning = MutableStateFlow(false)
    val relayWarning: StateFlow<Boolean> = _relayWarning.asStateFlow()

    /** Connected, but without everything it was supposed to carry. */
    private val _degraded = MutableStateFlow<CallDegradation?>(null)
    val degraded: StateFlow<CallDegradation?> = _degraded.asStateFlow()

    var channel: CallChannel = CallChannel.AUDIO
        private set
    var contactName: String = ""
        private set
    var contactAvatarUrl: String? = null
        private set
    var visitor: VisitorProfile? = null
        private set

    private var invitationId: String? = null
    private var callSessionId: String? = null
    private var started = false

    fun begin(
        invitation: CallInvitation,
        contactName: String,
        contactAvatarUrl: String?,
        visitor: VisitorProfile?,
    ) {
        if (started) return
        started = true
        this.invitationId = invitation.id
        this.channel = invitation.kind
        this.contactName = contactName
        this.contactAvatarUrl = contactAvatarUrl
        this.visitor = visitor
        _cameraOn.value = invitation.kind == CallChannel.VIDEO

        viewModelScope.launch { waitForVisitor() }
        viewModelScope.launch {
            room.events.collect { event -> onRoomEvent(event) }
        }
    }

    /**
     * Polling rather than realtime, on purpose.
     *
     * The app has no realtime transport, an invitation lives for five minutes,
     * and a request every two seconds for at most that long is cheaper to
     * build and to reason about than a socket that exists for this one screen.
     */
    private suspend fun waitForVisitor() {
        val id = invitationId ?: return
        while (_phase.value == CallPhase.Waiting) {
            runCatching { api.invitation(id) }.onSuccess { invitation ->
                val sessionId = invitation.callSessionId
                if (sessionId != null && invitation.isJoined) {
                    callSessionId = sessionId
                    join(sessionId)
                    return
                }
                if (invitation.isTerminal) {
                    _phase.value = CallPhase.Ended(
                        if (invitation.status == "declined") {
                            CallOutcome.Declined
                        } else {
                            CallOutcome.Expired
                        }
                    )
                    return
                }
            }
            // A single failed poll is a blip, not an answer. Only a terminal
            // invitation or the operator ends the wait.
            delay(POLL_MILLIS)
        }
    }

    private suspend fun join(callSessionId: String) {
        _phase.value = CallPhase.Connecting

        val credentials = runCatching { api.callToken(callSessionId, displayName = null) }
            .getOrElse {
                _phase.value = CallPhase.Ended(CallOutcome.Failed(it.message ?: "token"))
                return
            }
        _relayWarning.value = credentials.warnings?.contains("turn_missing") == true

        val url = credentials.signallingUrl
        if (url.isNullOrEmpty()) {
            _phase.value = CallPhase.Ended(CallOutcome.Failed("no_server_url"))
            return
        }

        val result = room.connect(url, credentials, wantsVideo = channel == CallChannel.VIDEO)
        when (result) {
            is CallRoom.Result.Failed -> {
                _phase.value = CallPhase.Ended(CallOutcome.Failed(result.reason))
                return
            }
            is CallRoom.Result.Joined -> {
                // Publishing is best-effort, deliberately. A camera that will
                // not start — none at all on an emulator, permission refused,
                // another app holding it — is a reason to carry on without
                // video, not a reason to drop a call the visitor has already
                // answered.
                if (!result.microphone) {
                    _muted.value = true
                    _degraded.value = CallDegradation.NO_MICROPHONE
                }
                if (channel == CallChannel.VIDEO && !result.camera) {
                    _cameraOn.value = false
                    if (_degraded.value == null) _degraded.value = CallDegradation.NO_CAMERA
                }
                room.setSpeaker(_speakerOn.value)
                _phase.value = CallPhase.Connected
                _connectedAt.value = Instant.now()
            }
        }
    }

    private fun onRoomEvent(event: CallRoom.Event) {
        when (event) {
            is CallRoom.Event.VisitorPresenceChanged -> {
                _visitorPresent.value = event.present
                // The visitor leaving the room IS the call ending. Waiting for
                // the server to notice would leave the operator looking at a
                // live-looking screen with nobody on the other end.
                if (!event.present && _phase.value == CallPhase.Connected) {
                    finish(CallOutcome.VisitorLeft)
                }
            }
            is CallRoom.Event.Disconnected -> {
                if (_phase.value.isLive) finish(CallOutcome.VisitorLeft)
            }
        }
    }

    // MARK: - Controls

    fun toggleMute() {
        val next = !_muted.value
        _muted.value = next
        viewModelScope.launch { room.setMicrophone(enabled = !next) }
    }

    fun toggleCamera() {
        if (channel != CallChannel.VIDEO) return
        val next = !_cameraOn.value
        _cameraOn.value = next
        viewModelScope.launch { room.setCamera(enabled = next) }
    }

    fun toggleSpeaker() {
        val next = !_speakerOn.value
        _speakerOn.value = next
        room.setSpeaker(next)
    }

    /** Ends the call for both sides. */
    fun hangUp() = finish(CallOutcome.HungUp)

    private fun finish(outcome: CallOutcome) {
        if (!_phase.value.isLive) return
        _phase.value = CallPhase.Ended(outcome)
        val sessionId = callSessionId
        viewModelScope.launch {
            room.disconnect()
            // Told to the server last and best-effort: the local side is
            // already over, and a failed hang-up request must not leave the
            // operator staring at a call they have finished with.
            if (sessionId != null) runCatching { api.hangUp(sessionId) }
        }
    }

    override fun onCleared() {
        // Whatever happened, the microphone and camera go back. A room left
        // open is a device the next app to ask for it is told no about.
        room.release()
        super.onCleared()
    }

    private companion object {
        const val POLL_MILLIS = 2_000L
    }
}
